crates/bevy_core_pipeline/src/prepass/node.rs

```diff
@@ -0,0 +1,133 @@
+};
+#[cfg(feature = "trace")]
+use bevy_utils::tracing::info_span;
+
+use super::{AlphaMask3dPrepass, Opaque3dPrepass, ViewPrepassTextures};
+
+/// Render node used by the prepass.
+///
+/// By default, inserted before the main pass in the render graph.
+pub struct PrepassNode {
+    main_view_query: QueryState<
+        (
+            &'static ExtractedCamera,
+            &'static RenderPhase<Opaque3dPrepass>,
+            &'static RenderPhase<AlphaMask3dPrepass>,
+            &'static ViewDepthTexture,
+            &'static ViewPrepassTextures,
+        ),
+        With<ExtractedView>,
+    >,
+}
+
+impl PrepassNode {
+    pub const IN_VIEW: &'static str = "view";
+
+    pub fn new(world: &mut World) -> Self {
+        Self {
+            main_view_query: QueryState::new(world),
+        }
+    }
+}
+
+impl Node for PrepassNode {
+    fn input(&self) -> Vec<SlotInfo> {
+        vec![SlotInfo::new(Self::IN_VIEW, SlotType::Entity)]
+    }
+
+    fn update(&mut self, world: &mut World) {
+        self.main_view_query.update_archetypes(world);
+    }
+
+    fn run(
+        &self,
+        graph: &mut RenderGraphContext,
+        render_context: &mut RenderContext,
+        world: &World,
+    ) -> Result<(), NodeRunError> {
+        let view_entity = graph.get_input_entity(Self::IN_VIEW)?;
+        let Ok((
+            camera,
+            opaque_prepass_phase,
+            alpha_mask_prepass_phase,
+            view_depth_texture,
+            view_prepass_textures,
+        )) = self.main_view_query.get_manual(world, view_entity) else {
+            return Ok(());
+        };
+
+        if opaque_prepass_phase.items.is_empty() && alpha_mask_prepass_phase.items.is_empty() {
+            return Ok(());
+        }
+
+        let mut color_attachments = vec![];
+        if let Some(view_normals_texture) = &view_prepass_textures.normal {
+            color_attachments.push(Some(RenderPassColorAttachment {
+                view: &view_normals_texture.default_view,
+                resolve_target: None,
+                ops: Operations {
+                    load: LoadOp::Clear(Color::BLACK.into()),
+                    store: true,
+                },
+            }));
+        }
+
+        {
+            // Set up the pass descriptor with the depth attachment and optional color attachments
+            let mut render_pass = render_context.begin_tracked_render_pass(RenderPassDescriptor {
+                label: Some("prepass"),
+                color_attachments: &color_attachments,
+                depth_stencil_attachment: Some(RenderPassDepthStencilAttachment {
+                    view: &view_depth_texture.view,
+                    depth_ops: Some(Operations {
+                        load: LoadOp::Clear(0.0),
+                        store: true,
+                    }),
+                    stencil_ops: None,
+                }),
+            });
+            if let Some(viewport) = camera.viewport.as_ref() {
```

Comment:
I don't think we need to set the viewport - the textures should always be created to match the viewport size, anyways.

Also, there's a RenderPhase::render() helper we can use that abstracts some of the boilerplate rendering in this file.

Reply:
Are you looking at an old branch? I'm not sure what boilerplate you are talking about otherwise, this is already using the RenderPhase::render function. For the viewport, I kept it there because it was copied from the main pass.

Reply:
Ah sorry, I was looking at an old branch. 

For the viewport, it's needed in the main pass, because you might be drawing to a texture not the same size as the viewport. For the prepass, we create the textures to match the viewport size to begin with. We don't need to handle the viewport in it.

Reply:
Yeah, that makes sense, I'll remove it.